cmd/sidecar/adapter/main.go

```diff
@@ -0,0 +1,142 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/n3wscott/sources/pkg/apis/sources/v1alpha1"
+	ceclient "github.com/n3wscott/sources/pkg/cloudeventclient"
+)
+
+type envConfig struct {
+	// Source options
+	Sink         string                    `envconfig:"K_SINK" required:"true"`
+	OutputFormat v1alpha1.OutputFormatType `envconfig:"K_OUTPUT_FORMAT" required:"true"`
+	Source       string                    `envconfig:"EVENT_SOURCE" required:"true"`
+	Type         string                    `envconfig:"EVENT_TYPE" required:"true"`
+
+	// Receiving options
+	Port        string `envconfig:"PORT" required:"true"`
+	ServePublic bool   `envconfig:"SERVE_PUBLICLY" default:"false"`
+}
+
+func makeReceive(ceclient cloudevents.Client, eventtype string, eventsrc cloudevents.URLRef) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "POST" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		id := r.URL.Path[len("/"):]
+
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Println("Could not read POST body:", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		event := cloudevents.Event{
+			Context: cloudevents.EventContextV02{
+				Type:   eventtype,
+				Source: eventsrc,
+				ID:     id, // will default if it's empty string
+			}.AsV02(),
+			Data: data,
+		}
+
+		log.Printf("Sending event with %d bytes of data\n", len(data))
+		resp, err := ceclient.Send(r.Context(), event)
+		if err != nil {
+			log.Println("Failed to send cloud event:", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		if resp == nil {
+			// No response from sink
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+
+		respBytes, err := resp.DataBytes()
+		if err != nil {
+			log.Println("Failed to read response:", err)
+			// Assume success so that event is not duplicated
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(respBytes)
+	}
+}
+
+func main() {
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Fatal("Failed to process env: ", err)
+	}
+
+	ceclient, err := ceclient.New(env.OutputFormat, env.Sink)
+	if err != nil {
+		log.Fatal("Could not create CloudEvents client: ", err)
+	}
+
+	shutdownChan := make(chan os.Signal, 1)
```

Comment:
an alternative to making your own signal is to use the `"knative.dev/pkg/signals"` pkg, like https://github.com/knative/eventing/blob/be4cd2ed42c18dd13229c809d2338d573886057b/vendor/knative.dev/pkg/injection/sharedmain/main.go#L93

And then you can still block on `ctx.Done()`